### DISCUSSION TITLE
Proposal: split up all code shared between pipelines.

### DISCUSSION THREAD

**Discussion by mintlu8** [bevyengine/bevy#13770] [Ideas]
# Why

The pipelined rendering system that bevy currently have is great but feels almost impossible to learn. This is due to the fact a lot of code is shared between multiple pipelines, gated behind if statements and wgsl `#ifdef`s, making them difficult to read or modify.

# Proposed Solution

* Remove all wgsl definitions regarding pipelines, split up shader code if needed.
* Split up `specialize` on `Material` into 3 functions for each pipeline.
  * `specialize`
  * `prepass_specialize`
  * `deferred_specialize`

**Comment by aloucks** [comment]
I ran into this as well. It's quite difficult to grok what's going on with the sheer volume of `#ifdef`s. The issue is exasperated by the lack of indentation in the preprocessor statements and sometime incorrect indentation of the WGSL code itself. Looking at the `naga_oil` crate, it supports a version of function polymorphism (e.g. virtual functions). I'm wondering if leveraging that feature could make things more manageable. This might have an added bonus of making customizing aspects of the shaders easier.
